scylla/src/transport/load_balancing/mod.rs

```diff
@@ -25,13 +33,32 @@ pub struct Statement<'a> {
     pub is_confirmed_lwt: bool,
 }
 
-pub type Plan<'a> = Box<dyn Iterator<Item = Arc<Node>> + Send + Sync + 'a>;
+pub type FallbackPlan<'a> = Box<dyn Iterator<Item = NodeRef<'a>> + Send + Sync + 'a>;
 
-/// Policy that decides which nodes to contact for each query
+/// Policy that decides which nodes to contact for each query.
+///
+/// When a query is prepared to be sent to ScyllaDB/Cassandra, a `LoadBalancingPolicy`
+/// implementation constructs a load balancing plan. That plan is a list of nodes to which
+/// the driver will try to send the query. The first elements of the plan are the nodes which are
+/// the best to contact (e.g. they might have the lowest latency).
+///
+/// Most queries are send on the first try, so the query execution layer rarely needs to know more
+/// than one node from plan. To better optimize that case, `LoadBalancingPolicy` has two methods:
+/// `pick` and `fallback`. `pick` returns a first node to contact for a given query, `fallback`
+/// returns the rest of the load balancing plan.
+///
+/// `fallback` is called only after a failed send to `pick`ed node (or when executing
+/// speculatively).
+/// If a `pick` returns `None`, `fallback` will not be called.
+///
+/// Usually the driver needs only the first node from load balancing plan (most queries are send
+/// successfully, and there is no need to retry).
+///
+/// This trait is used to produce an iterator of nodes to contact for a given query.
 pub trait LoadBalancingPolicy: Send + Sync + std::fmt::Debug {
-    /// It is used for each query to find which nodes to query first
-    fn plan<'a>(&self, statement: &Statement, cluster: &'a ClusterData) -> Plan<'a>;
+    /// Returns the first node to contact for a given query.
+    fn pick<'a>(&'a self, query: &'a QueryInfo, cluster: &'a ClusterData) -> Option<NodeRef<'a>>;
 
-    /// Returns name of load balancing policy
-    fn name(&self) -> String;
+    /// Returns all contact-appropriate nodes for a given query.
+    fn fallback<'a>(&'a self, query: &'a QueryInfo, cluster: &'a ClusterData) -> FallbackPlan<'a>;
```

Comment:
I'm sorry, but I'm in favor of not making such gradual changes in this pull request development state. I feel it would add a lot of time and the benefit would be rather small.